### PR TITLE
feat: add paddle image encoder

### PR DESCRIPTION
**Pull request by Yongxuanzhang**
This PR migrates the paddle image encoder to 2.0 
https://github.com/jina-ai/internal-tasks/issues/92

### CODE REVIEW

File: encoder_paddle.py [hunk 1]
```diff
@@ -1,140 +1,96 @@
...
+class ImagePaddlehubEncoder(Executor):
     def __init__(
         self,
-        target_size: Union[Iterable[int], int] = 224,
-        img_mean: Tuple[float] = (0, 0, 0),
-        img_std: Tuple[float] = (1, 1, 1),
-        resize_dim: Union[Iterable[int], int] = 256,
-        channel_axis: int = -1,
-        target_channel_axis: int = -1,
+        model_name: str = None,
```
Comment (JoanFM) [inline encoder_paddle.py]: proper typehint, use Optional

**Review by JoanFM — COMMENTED** [review]
The pooling strategy needs to be revisited

File: paddle_image.py [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+    :param default_traversal_path: traversal path of the Documents, (e.g. 'r', 'c')
+    :param on_gpu: set to True if using GPU
+    :param args:  Additional positional arguments
+    :param kwargs: Additional keyword arguments
+    """
+    def __init__(
+            self,
+            model_name: Optional[str] = None,
+            pool_strategy: Optional[str] = None,
+            channel_axis: int = -3,
```
Comment (JoanFM) [inline paddle_image.py:55]: I think channel axis should not be here as for clip

**Review by JoanFM — COMMENTED** [review]
check the pooling strategy as commented on slack

**Review by JoanFM — DISMISSED** [review]
Please have both unit and integration tests

**Comment by Yongxuanzhang** [comment]
Yes I have both tests > Please have both unit and integration tests

**Review by JoanFM — COMMENTED** [review]
Ok, and what about pooling, how do Paddle do this?

File: README.md [hunk 3]
```diff
@@ -1,19 +1,23 @@
...
+    def process_response(resp):
+        ...
+
+    f = Flow().add(uses={
+        'jtype': ImagePaddlehubEncoder.__name__,
+        'with': {
+            'default_batch_size': 32,
+            'model_name': 'xception71_imagenet',
+        }
+    })
```
Comment (numb3r3) [inline README.md]: I think you forget the `metas`

```suggestion
    f = Flow().add(uses={
        'jtype': ImagePaddlehubEncoder.__name__,
        'with': {
            'default_batch_size': 32,
            'model_name': 'xception71_imagenet',
        },
        'metas': {
            'py_modules': ['paddle_image.py']
        }
    })
```

**Review by numb3r3 — DISMISSED** [review]

File: README.md (hunk 3 above)
Comment (Yongxuanzhang) [reply README.md]: thanks!

File: manifest.yml [hunk 4]
```diff
@@ -1,12 +1,10 @@
 manifest_version: 1
-name: ImageNormalizer
-kind: crafter
-type: pod
-description: '`ImageNormalizer` resizes and normalizes images'
+name: ImagePaddlehubEncoder
+kind: encoder
```
Comment (cristianmtr) [inline manifest.yml]: `kind` doesn't exist in the new manifest spec https://docs.google.com/document/d/1mLNkLlg4aiwhHmM0Nq-ndzbI8cTUQH0Ub8qgtnPqsFI/edit#heading=h.cz2166romzp

File: tests/integration/test_flow_integration.py [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+    def validate_traversal(expected_docs_per_path: List[List[str]]):
+        def validate(resp):
+            for path, count in expected_docs_per_path:
+                assert len(DocumentArray(resp.data.docs).traverse_flat([path]).get_attributes('embedding')) == count
+
+        return validate
+
+    flow = Flow().add(uses=ImagePaddlehubEncoder)
+    with flow:
+        flow.post(
```
Comment (cristianmtr) [inline tests/integration/test_flow_integration.py]: shouldn't you assert the shape (at least) of the encoded docs?

File: tests/unit/test_encoder.py [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+
+def test_imagepaddlehubencoder_encode(test_images: Dict[str, np.array]):
+    encoder = ImagePaddlehubEncoder(channel_axis=3)
+
+    embeddings = {}
+    for name, image_arr in test_images.items():
+        docs = DocumentArray([Document(blob=image_arr)])
+        encoder.encode(docs, parameters={})
+        embeddings[name] = docs[0].embedding
+    print(embeddings)
```
Comment (cristianmtr) [inline tests/unit/test_encoder.py]: remove print. It's not helpful, I think

**Review by cristianmtr — CHANGES_REQUESTED** [review]

File: paddle_image.py (hunk 2 above)
Comment (Yongxuanzhang) [reply paddle_image.py:55]: It should be ok to keep this maybe? I know we should do this in normalizer

File: tests/unit/test_encoder.py (hunk 6 above)
Comment (Yongxuanzhang) [reply tests/unit/test_encoder.py]: thanks! Just forgot to remove

File: tests/integration/test_flow_integration.py (hunk 5 above)
Comment (Yongxuanzhang) [reply tests/integration/test_flow_integration.py]: sure!

**Comment by Yongxuanzhang** [comment]
> Ok, and what about pooling, how do Paddle do this?

I think by default we should continue to use mean pooling, if not use pooling, then the embedding would be (2048, 1,1), the mean pooling won't change the value but instead flat the (2048,1,1) into (2048,). Do you think this is correct or not? @JoanFM

**Review by JoanFM — COMMENTED** [review]
Ok, just keep on the top of our heads this in mind

**Comment by slettner** [comment]
Please add `.dockerignore` 

```
.git
.venv
.github
.pytest_cache
tests
__pycache__
scripts
```

File: README.md [hunk 7]
```diff
@@ -1,19 +1,26 @@
...
+            'default_batch_size': 32,
+            'model_name': 'xception71_imagenet',
+        },
+        'metas': {
+            'py_modules': ['paddle_image.py']
+        }
+    })
+    with f:
+        f.post(on='/test', inputs=(Document(blob=np.ones((224, 224, 3))) for _ in range(25)), on_done=process_response)
+```
```
Comment (slettner) [inline README.md:102]: I find these two section very useful when looking at executors (just an example from https://github.com/jina-ai/executor-template): 

```
### Inputs 

`Document` with `blob` of the shape `256`.

### Returns

`Document` with `embedding` fields filled with an `ndarray` of the shape `embedding_dim` (=128, by default) with `dtype=nfloat32`.
```

Maybe we can add this here as well?

File: paddle_image.py [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+        module = hub.Module(name=self.model_name)
+        inputs, outputs, self.model = module.context(trainable=False)
+        self._get_inputs_and_outputs_name(inputs, outputs)
+
+        import paddle.fluid as fluid
+        self.device = fluid.CUDAPlace(0) if self.on_gpu else fluid.CPUPlace()
+        self.exe = fluid.Executor(self.device)
+
+    @requests
+    def encode(self, docs: DocumentArray, parameters: dict, **kwargs) -> DocumentArray:
```
Comment (slettner) [inline paddle_image.py]: Function returns `None` actually

Comment (JoanFM) [reply paddle_image.py]: yes, it should not be anotated with this `return`

File: paddle_image.py [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+        filtered_docs = [doc for doc in flat_docs if doc.blob is not None]
+
+        return _batch_generator(filtered_docs, batch_size)
+
+    def _get_pooling(self, content: 'np.ndarray') -> 'np.ndarray':
+        """Get ndarray with selected pooling strategy"""
+        _reduce_axis = tuple((i for i in range(len(content.shape)) if i > 1))
+        return getattr(np, self.pool_strategy)(content, axis=_reduce_axis)
+
+    def _get_inputs_and_outputs_name(self, input_dict, output_dict):
```
Comment (slettner) [inline paddle_image.py:126]: I would prefer a functional approach that sets `self.inputs_name` and `self.outputs_name` directly in the constructor, i.e. 
```
self.inputs_name, self.outputs_name = self._get_inputs_and_outputs_name(input, output)
```

Comment (JoanFM) [reply paddle_image.py:126]: Yes, even more if u call it as `_getter`. It is much easier to read if the variables are explicitly initialized in `__init__`

**Review by slettner — COMMENTED** [review]
Left some comments on minor things, great work!

File: Dockerfile [hunk 10]
```diff
@@ -1,7 +1,14 @@
-FROM jinaai/jina:master
+FROM jinaai/jina:2.0.0rc6
```
Comment (florian-hoenicke) [inline Dockerfile]: GPU support needs to be added. You can pair with @bwanglzu. He implemented the same with text.

Comment (JoanFM) [reply Dockerfile]: How are we testing this?

File: paddle_image.py (hunk 8 above)
Comment (Yongxuanzhang) [reply paddle_image.py]: good  catch!

File: paddle_image.py (hunk 9 above)
Comment (Yongxuanzhang) [reply paddle_image.py:126]: good point

File: README.md (hunk 7 above)
Comment (Yongxuanzhang) [reply README.md:102]: Thanks! I will update readme according to the template

**Comment by Yongxuanzhang** [comment]
> Please add `.dockerignore`
> 
> ```
> .git
> .venv
> .github
> .pytest_cache
> tests
> __pycache__
> scripts
> ```

added

File: Dockerfile [hunk 11]
```diff
@@ -1,7 +1,14 @@
-FROM jinaai/jina:master
+FROM jinaai/jina:2.0.0rc8
 
-ADD *.py *.yml requirements.txt ./
+# install git
+RUN apt-get -y update && apt-get install -y git
 
+# install requirements before copying the workspace
+COPY requirements.txt /requirements.txt
```
Comment (cristianmtr) [inline Dockerfile:7]: Why are we copying the requirements before the workspace?

Comment (Yongxuanzhang) [reply Dockerfile:7]: I copy this from some template. How about this? 
image

**Review by cristianmtr — APPROVED** [review]

**Pull request merged**